### PR TITLE
WIP - Feature: Custom OID polling and graphing

### PR DESCRIPTION
**Pull request by slashdoom**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

### CODE REVIEW

**Comment by slashdoom** [comment]
This is just an attempt to expand upon @erotel 's PR #7925 that I've been playing with.  I really don't know if I'm on the right track or not so decided to PR for review and comment.

Please note sql schema file is 999.sql as I didn't know how these get managed.

**Comment by strehi** [comment]
Tested it against v1.42.01-69-gd46f300 - it's looking promising good.
How do we get this upstream?
TIA

File: sql-schema/999.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+CREATE TABLE `customoids` ( `customoid_id` int(11) NOT NULL AUTO_INCREMENT PRIMARY KEY, `device_id` int(11) NOT NULL DEFAULT '0', `customoid_descr` varchar(255) NULL DEFAULT '', `customoid_deleted` tinyint(1) NOT NULL DEFAULT '0', `customoid_current` double NULL, `customoid_prev` double NULL, `customoid_oid` varchar(255) NULL, `customoid_unit` varchar(20) NULL, `customoid_divisor` int(11) NOT NULL DEFAULT '1', `customoid_multiplier` int(11) NOT NULL DEFAULT '1', `customoid_limit` double NULL, `customoid_limit_warn` double NULL, `customoid_limit_low` double NULL, `customoid_limit_low_warn` double NULL, `customoid_alert` tinyint(1) NOT NULL DEFAULT '0', `customoid_passed` tinyint(1) NOT NULL DEFAULT '0', `lastupdate` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP, `user_func` varchar(100) NULL );
```
Comment (murrant) [inline sql-schema/999.sql]: Setting this to 999 will cause many issues.  For one, users who have tested this without reverting their database by hand will not get schema updates until we reach schema 1000...

**Comment by slashdoom** [comment]
Is there a proper way to handle this?    The PR is about 3 months old now.  I did mention the 999 file in the comments but it seems like any normal number I chose would have given testers problems as well?

**Comment by LoZio** [comment]
+1 on this, would be very useful

**Comment by kalamchi75** [comment]
Hi Guys, 

Following all the 19 files changes, the feature is available in my LibreNMS. 
I get the custom OID page and testing the OID works and retrieves the value. 
However it fails to save the OID. Please see below 
fail-to-save

Any solution please ?

Thanks

**Comment by slashdoom** [comment]
Does the customoids table exist in your librenms DB?  If not you might need to run the apply again and run ./daily.sh right after to get your SQL schema to update.

**Comment by kalamchi75** [comment]
Thanks slashdoom. 
I have rebooted the server and checked the DB. I can see the table customoid there. 
I have tried again to add few OIDs, and successfully saved them. Awesome. 
Two of those OIDs are already graphing, but two still produce -NAN in the graph, but I can see the values in the Custom OID edit page. 
see below: 
customeoid-graphs
customeoid-settings

The rrdtool output seems ok. No errors there. i'm not sure where to look. 

Also one more thing, would be possible to add an option for the graphing color ?

Many thanks

**Comment by kalamchi75** [comment]
Hi, 

OK I found why those graphs were not being updated. The values retrieved are negative, and it seems that the rrdtool with customoid does not scale to negative values. Once I multiplied the values by -1, making them positive, the graphs started to be updated. 
I would kindly ask if this can be addressed and fixed. 

Thanks

**Comment by kalamchi75** [comment]
Hi Slashdoom,

The fixes work. I will paste some sample graphs.
I think it’s an awesome feature and really helps when monitoring non-standard equipment with sensors that are not auto-detected by LibreNMS. We have a bunch of them.
I am trying to get LibreNMS to monitor the different type of device we use in our job, and eventually replace the Cacti servers we have been using.

The color feature would be also awesome if you can add it.

Many thanks

-------------------------
Ali Kalamchi,
Senior System Engineer,
[cid:image001.png@01D07862.D24D6500]
Communicate without boundaries

Talia Deutschland GmbH
Mitterweg 36
82399 Raisting
Germany

T:  +49 (0)8807 3692223
F: +49 (0)8807 2140036
E: ali.kalamchi@talia.net<mailto:ali.kalamchi@talia.net>
www.talia.net<http://www.talia.net/>

**Comment by kalamchi75** [comment]
Below screen shots of the fixed customoid's graphs. 
fixed-graphs

I really like this feature. Kudos

**Comment by angelfoga** [comment]
Is the use of this is to get oids that can't get discovered via yaml (devices with bad snmp implementations)? 
In my case this means that the poller will mark them as down, `status_reason= 'snmp'` and will not get the current values
image

**Comment by slashdoom** [comment]
Thanks for testing.  This could certainly be used to poll OID's not discoverable by the yaml definitions but it does currently require that the device pass the poller's device_is_up() test.  My thinking is that this is somewhat for the protection of the poller itself.  If the test is bypassed for the customoid poller module but the device really is down it could end up hurting poller performance overall.

**Comment by angelfoga** [comment]
> We have also some specific devices that has SNMP working, but they are not common and not added in the auto-detection of LibreNMS, something Comtech HPOD and Comtech LPOD satellite power amplifiers.

Yeah, I maybe I didn't explain myself. My company has some FM Transmitters that doesn't discover either, **and this solved my problem.**

> I haven't tested the feature with devices that poll only with ping. I mean if they don't have proper SNMP implementation then I guess one won't get any values anyway, whether with LibreNMS or any other SNMP monitoring (cacti, whatsup gold, ... etc)

With nagios I was able to get all the values because they're basically snmpgets showing a string. No one is going to complain if I can't monitor 100% these devices, their snmp is screwed as hell... so yeah, we're on the same page on this ;)

**Comment by jozefrebjak** [comment]
Thank you for this. It is working perfectly for me. I just have one feature request. Will be possible to add option to choose between GAUGE and COUNTER syntax for graphing ? Thank you.

**Comment by slashdoom** [comment]
Good idea, a GAUGE / COUNTER selector should be included now.  A couple caveats...

1. It's only selectable when adding new OID's; you can't edit to change existing ones since afaik that would require the rrd files being recreated in most cases.
 
2. Along those same lines, you'll need to manually delete existing rrd files that you want to change to counters.
 
3. It requires a schema update.  If you don't want to lose your existing OID's when you apply this PR you can run the MySQL command:
 ```
 ALTER TABLE `customoids`
 ADD COLUMN `customoid_datatype` varchar(20) NOT NULL DEFAULT 'GAUGE' AFTER `customoid_oid`;
 ```

**Comment by slashdoom** [comment]
Great, thanks for testing!

**Comment by laf** [comment]
I do think this is a great thing to include but in my opinion it's the one thing we have to get right first go rather than iterate through versions to get it right. This is what I'd like to see in it:

Add a 'type' so we could group them within the webui.

Add the ability to create 'global' custom OIDs so they can be re-used across devices.

Add the ability to use yaml so we can have a custom dir for OS where people could create the global OIDs as well. I.e. support within `includes/definitions/discovery/*.yaml` + also `includes/definitions/discovery/custom/*.yaml` which we could merge in.

Add the ability to skip the tests for custom OIDs. Distributed polling means the webui may not have permission to query via snmp only via the pollers.

**Comment by murrant** [comment]
A button that says "Contribute this sensor to LibreNMS" that auto opens a PR on github. :smile:

**Comment by ArtemKabanov** [comment]
It's very good feature! I'm waiting this very match.

**Comment by slashdoom** [comment]
Thanks all.  Hoping to take this feedback and have another look at this after the holidays.

**Comment by slashdoom** [comment]
Updated schema file so it should hopefully patch.  If it works, look for the Custom OID option under Edit for your device to add new OIDs (see angelfoga's screenshot).  After a couple polls a Custom OID tab should also show in your device's Graphs pane.

**Comment by airkingbr** [comment]
I have a problem installing this PR.

root@noc:/opt/librenms# ./scripts/github-apply 8738
<stdin>:811: trailing whitespace.
 
<stdin>:841: trailing whitespace.
        
<stdin>:987: trailing whitespace.
    
Checking patch html/includes/forms/customoid.inc.php...
Checking patch html/includes/forms/delete-customoid.inc.php...
Checking patch html/includes/forms/parse-customoid.inc.php...
Checking patch html/includes/graphs/customoid/auth.inc.php...
Checking patch html/includes/graphs/customoid/customoid.inc.php...
Checking patch html/includes/graphs/graph.inc.php...
Checking patch html/includes/modal/delete_customoid.inc.php...
Checking patch html/includes/modal/new_customoid.inc.php...
Checking patch html/includes/print-customoid.php...
Checking patch html/pages/device/edit.inc.php...
Checking patch html/pages/device/edit/customoid.inc.php...
Checking patch html/pages/device/graphs.inc.php...
Checking patch html/pages/device/graphs/customoid.inc.php...
Checking patch includes/common.php...
Hunk #1 succeeded at 973 (offset 7 lines).
Hunk #2 succeeded at 1675 (offset -4 lines).
Checking patch includes/defaults.inc.php...
Hunk #1 succeeded at 794 (offset 2 lines).
Checking patch includes/definitions.inc.php...
error: while searching for:
$config['graph_types']['device']['pf_memdropped']['order']           = '10';
$config['graph_types']['device']['pf_memdropped']['descr']           = 'MemDropped';





//Pulse Secure Graphs
$config['graph_types']['device']['pulse_users']['section']         = 'firewall';
$config['graph_types']['device']['pulse_users']['order']           = '0';

error: patch failed: includes/definitions.inc.php:279
error: includes/definitions.inc.php: patch does not apply
Checking patch includes/polling/customoid.inc.php...
Checking patch misc/alert_rules.json...
Checking patch sql-schema/276.sql...

**Comment by airkingbr** [comment]
Have some updates on definitions.inc.php , because that tue patch fail to apply.
The workaround is downgrade the file definitions.inc.php to past version and patch does the magic.

Thanks.
AirKing

**Comment by murrant** [comment]
@slashdoom You need to change to a migration (this will keep you from having to update it all the time) https://laravel.com/docs/5.4/migrations

**Comment by PipoCanaja** [comment]
@airkingbr : you can have a look at https://github.com/librenms/librenms/pull/9592/files#diff-468c1431c137e67fda6e221cc38ad243 to see how to create a migration file. 
Then build-schema.php is still needed. 
No more need to change the number.  But still necessary to rebase at the end of the PR, before the merge, and re-run build-schema.php.

**Comment by adman10** [comment]
Hi - this feature looks really useful for my environment to poll some custom OID's for VPN clients on a strongswan install.

I'm using SNMP extend to make these available to poll

I'm getting the below error that @airkingbr mentions when trying to install this PR.

@slashdoom - any suggestions?

<stdin>:811: trailing whitespace.

<stdin>:841: trailing whitespace.

<stdin>:987: trailing whitespace.

Checking patch html/includes/forms/customoid.inc.php...
Checking patch html/includes/forms/delete-customoid.inc.php...
Checking patch html/includes/forms/parse-customoid.inc.php...
Checking patch html/includes/graphs/customoid/auth.inc.php...
Checking patch html/includes/graphs/customoid/customoid.inc.php...
Checking patch html/includes/graphs/graph.inc.php...
Checking patch html/includes/modal/delete_customoid.inc.php...
Checking patch html/includes/modal/new_customoid.inc.php...
Checking patch html/includes/print-customoid.php...
Checking patch html/pages/device/edit.inc.php...
Checking patch html/pages/device/edit/customoid.inc.php...
Checking patch html/pages/device/graphs.inc.php...
Checking patch html/pages/device/graphs/customoid.inc.php...
Checking patch includes/common.php...
Hunk #1 succeeded at 958 (offset -8 lines).
Hunk #2 succeeded at 1667 (offset -12 lines).
Checking patch includes/defaults.inc.php...
Hunk #1 succeeded at 794 (offset 2 lines).
Checking patch includes/definitions.inc.php...
error: while searching for:
$config['graph_types']['device']['pf_memdropped']['order']           = '10';
$config['graph_types']['device']['pf_memdropped']['descr']           = 'MemDropped';





//Pulse Secure Graphs
$config['graph_types']['device']['pulse_users']['section']         = 'firewall';
$config['graph_types']['device']['pulse_users']['order']           = '0';

error: patch failed: includes/definitions.inc.php:279
error: includes/definitions.inc.php: patch does not apply
Checking patch includes/polling/customoid.inc.php...
Checking patch misc/alert_rules.json...
error: while searching for:
    "name": "Wireless Sensor under limit",
    "default": true
  },
  {
    "rule": "macros.state_sensor_critical",
    "name": "State Sensor Critical",

error: patch failed: misc/alert_rules.json:58
error: misc/alert_rules.json: patch does not apply
Checking patch sql-schema/282.sql...

**Comment by slashdoom** [comment]
I would hold off on using it right now.  I'm putting it back in WIP status until I can get the new SQL schema stuff figured out and try to work in some of the feedback.

**Comment by ibarrere** [comment]
This feature would definitely be useful for me. This is basically the one reason we're still on Observium.

**Comment by jsteinlechner** [comment]
thanks for this, although it's not working as expected. test is ok, but if i'm trying to save the sensor, it replies with an error "failed to add"... nothing more....

help appreciated!
thx

**Comment by slashdoom** [comment]
Sorry all, I wouldn't expect much out of this right now.  I haven't given up on the PR but I haven't had time to figure out the changes needed for the new Laravel stuff or to make the requested changes to get the concept right.  Hopefully I'll get some time and be able to circle back to this soon...

**Comment by Bobdave** [comment]
Question, I successfully installed the plugin and it works properly, but I don't have any "customoid" things in the alert drop down list. I do have them listed in the alert collection but they aren't doing anything if I select them... 

image

image

Everything else related to the customoid is working properly. We do have customoid saved and we are polling them and having good graphs

**Comment by hidagar** [comment]
Hi,

when I try to install I get this error:
root@librenms:/opt/librenms# ./scripts/github-apply 8738
<stdin>:811: trailing whitespace.

<stdin>:841: trailing whitespace.

<stdin>:987: trailing whitespace.

Checking patch html/includes/forms/customoid.inc.php...
Checking patch html/includes/forms/delete-customoid.inc.php...
Checking patch html/includes/forms/parse-customoid.inc.php...
Checking patch html/includes/graphs/customoid/auth.inc.php...
Checking patch html/includes/graphs/customoid/customoid.inc.php...
Checking patch html/includes/graphs/graph.inc.php...
error: html/includes/graphs/graph.inc.php: No such file or directory
Checking patch html/includes/modal/delete_customoid.inc.php...
Checking patch html/includes/modal/new_customoid.inc.php...
Checking patch html/includes/print-customoid.php...
Checking patch html/pages/device/edit.inc.php...
error: html/pages/device/edit.inc.php: No such file or directory
Checking patch html/pages/device/edit/customoid.inc.php...
Checking patch html/pages/device/graphs.inc.php...
error: html/pages/device/graphs.inc.php: No such file or directory
Checking patch html/pages/device/graphs/customoid.inc.php...
Checking patch includes/common.php...
Hunk #1 succeeded at 954 (offset -12 lines).
Hunk #2 succeeded at 1639 (offset -40 lines).
Checking patch includes/defaults.inc.php...
error: while searching for:
$config['poller_modules']['ntp']                         = true;
$config['poller_modules']['loadbalancers']               = false;
$config['poller_modules']['mef']                         = false;

// List of discovery modules. Need to be in this array to be
// considered for execution.

error: patch failed: includes/defaults.inc.php:792
error: includes/defaults.inc.php: patch does not apply
Checking patch includes/definitions.inc.php...
error: while searching for:
$config['graph_types']['device']['pf_memdropped']['order']           = '10';
$config['graph_types']['device']['pf_memdropped']['descr']           = 'MemDropp                                                                                                             ed';





//Pulse Secure Graphs
$config['graph_types']['device']['pulse_users']['section']         = 'firewall';
$config['graph_types']['device']['pulse_users']['order']           = '0';

error: patch failed: includes/definitions.inc.php:279
error: includes/definitions.inc.php: patch does not apply
Checking patch includes/polling/customoid.inc.php...
Checking patch misc/alert_rules.json...
error: while searching for:
    "name": "Wireless Sensor under limit",
    "default": true
  },
  {
    "rule": "macros.state_sensor_critical",
    "name": "State Sensor Critical",

error: patch failed: misc/alert_rules.json:58
error: misc/alert_rules.json: patch does not apply
Checking patch sql-schema/282.sql...

**Comment by louis-oui** [comment]
New pull request as this one has conflicts with master. Please review this change https://github.com/librenms/librenms/pull/10945

**Comment by louis-oui** [comment]
@PipoCanaja 
Pull request can be closed. Merged in https://github.com/librenms/librenms/pull/10945

**Pull request closed without merging**